### PR TITLE
Pkg resolution improvements: propagate requirements (with backtrace)

### PR DESCRIPTION
**Pull request by carlobaldassi**
This adds a pre-pruning step in which requirements are propagated along the dependency graph, keeping a backtrace of the process. The logic is explained [here](https://github.com/JuliaLang/julia/blob/41f9d6b3301d65a291870a096ce56e941599a4a7/base/pkg/query.jl#L266-L269): if all allowed versions of a required package require some other package, then compute the union of all the allowed versions for that other package, and treat that as a new requirement. At the end, versions outside the allowed ranges are pruned.

This has two benefits:
1. It makes things easier for the solver, by simplifying the graph and reducing the number of versions
2. It should detect most common problems with requirements that introduce implicit contradictions, and (since it keeps track of the process) it provides a useful error message to fix the situation.

Take for example the second issue reported in #20313. The previous code used to produce this:

```
julia> Pkg.pin("ColorTypes",v"0.3.0") ### the error message below is blaming the wrong package!!!
INFO: Creating ColorTypes branch pinned.a846a532.tmp
ERROR: unsatisfiable package requirements detected: no feasible version could be found for package: ImageRegistration
```
(the solver detected a problem, but the error message was unhelpful and actually misleading.)

With this PR the problem is detected before invoking the solver, and it produces this instead:

```
julia> Pkg.pin("ColorTypes", v"0.3.0")
INFO: Package ColorTypes: checking out existing branch pinned.a846a532.tmp
ERROR: Unsatisfiable requirements detected for package FixedPointNumbers:
├─version range [0.3.0,∞) required by package ColorTypes, whose only allowed version is 0.3.0:
│ └─version 0.3.0 set by fixed requirement (package is checked out, dirty or pinned)
├─version range [0.0.0-,∞) required by package ImageRegistration, whose allowed version range is [0.0.0-,∞):
│ └─version range [0.0.0-,∞) set by an explicit requirement
└─version range [0.0.0-,0.3.0) required by package Images, whose allowed version range is [0.0.0-,∞):
  └─version range [0.0.0-,∞) required by package ImageRegistration, whose allowed version range is [0.0.0-,∞):
    └─[see above for ImageRegistration backtrace]
The intersection of the requirements is empty.
```

cc @tkelman

I suppose that in the future some of this code could be reused and be relevant for addressing #13340 too.

Edit: I have also just changed the error message in case of failure of the solver, to at least improve the situation in cases like the one of #20313. Now it prints something like this:

```
ERROR: resolve is unable to satisfy package requirements.
  The problem was detected when trying to find a feasible version
  for package X.
  However, this only means that package X is involved in an
  unsatifiable or difficult dependency relation, and the root of
  the problem may be elsewhere.
```

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
Nice. I'm going to have to spend some time groking all of the resolution code so that we can refit it onto Pkg3 when it's ready for that.

**Comment by carlobaldassi** [comment]
I'm not sure about the status of Pkg3, but from what I had seen in the first proposal drafts adapting the code should be rather straightforward (actually, it would simplify quite a lot some things, since from what I understand we'd essentially only consider major.minor versions). If you have some code in the works I can definitely help in making the necessary changes.

**Comment by tkelman** [comment]
> we'd essentially only consider major.minor versions

That part of the proposal needs revision. There was discussion in the juleps issues, I think removing resolution's ability to distinguish patch versions would be harmful.

**Comment by carlobaldassi** [comment]
Ah I see, thanks. I suppose I should go and have a look. (I'm still willing to help regardless :) )

**Comment by carlobaldassi** [comment]
Objections to merge?

**Comment by StefanKarpinski** [comment]
No, go for it. Squash or not as appropriate.

**Comment by tkelman** [comment]
The end result looks like a huge improvement. Hadn't gotten to reviewing the implementation in detail, will do so when I get a chance.

**Comment by tkelman** [comment]
will have to check if those unicode line drawing characters are legible in the windows terminal

**Comment by carlobaldassi** [comment]
I remember those characters being used in [Microsoft Works for DOS](https://en.wikipedia.org/wiki/Microsoft_Works#/media/File:Ms-works-2.0-dos.png), I sincerely hope so.

**Comment by carlobaldassi** [comment]
I just checked (with julia 0.5), they do render correctly.

**Comment by tkelman** [comment]
cool. on what windows version?

**Comment by carlobaldassi** [comment]
7

File: base/pkg/types.jl [hunk 1]
```diff
@@ -40,15 +41,73 @@ show(io::IO, s::VersionSet) = join(io, s.intervals, " ∪ ")
 isempty(s::VersionSet) = all(isempty, s.intervals)
 in(v::VersionNumber, s::VersionSet) = any(i->in(v,i), s.intervals)
 function intersect(A::VersionSet, B::VersionSet)
-    ivals = vec([ intersect(a,b) for a in A.intervals, b in B.intervals ])
-    ivals = copy(ivals) # temporary bandaid for issue #4592
+    ivals = [intersect(a,b) for a in A.intervals for b in B.intervals]
```
Comment (tkelman) [inline base/pkg/types.jl]: I'm getting
```
ERROR: LoadError: LoadError: ArgumentError: argument to Flatten must contain at least one iterator
 in start at .\iterator.jl:599
 in grow_to! at .\array.jl:357
 in intersect at .\pkg\types.jl:44
 in push! at .\pkg\types.jl:182
 in requirements at .\pkg\query.jl:34
 in resolve at .\pkg\entry.jl:476
```
when I try to backport this to release-0.5 - the old version with `vec` should be equivalent right?

Comment (carlobaldassi) [inline base/pkg/types.jl]: Yes it's equivalent, this was just removing an outdated band-aid. Sorry, maybe next time I should put such things in a separate commit.

Comment (tkelman) [inline base/pkg/types.jl]: It's already fairly granular. Also note that I had to adjust an assert to avoid anonymous type parameter syntax, I think https://github.com/JuliaLang/julia/commit/f907b5716805013fcd65b1720d1ff940e11ed815#diff-cac60f612c90efb7c227b01575fb8e15R214 should be equivalent

Comment (carlobaldassi) [inline base/pkg/types.jl]: Yep, that should work fine too.
